File: nbodysolver-py/python/nbodysolver/nbodywrapper.py
import numpy as np
import numpy.typing as npt
from nbodysolver.nbodysolver import simulate_nbody_and_process

class NBodySystem:
    def __init__(self, 
                 r_list: npt.NDArray[np.float64], 
                 v_list: npt.NDArray[np.float64], 
                 m_list: npt.NDArray[np.float64]):
        self.r_list = r_list
        self.v_list = v_list
        self.m_list = m_list

class NBodyResults:
    def __init__(self, 
                 all_time: npt.NDArray[np.float64], 
                 all_r: npt.NDArray[np.float64], 
                 all_v: npt.NDArray[np.float64], 
                 all_ke: npt.NDArray[np.float64], 
                 all_pe: npt.NDArray[np.float64], 
                 all_te: npt.NDArray[np.float64]):
        self.all_time = all_time
        self.all_r = all_r
        self.all_v = all_v
        self.all_ke = all_ke
        self.all_pe = all_pe
        self.all_te = all_te
        
def simulate_system(system: NBodySystem, dt: float, 
                    max_time: float, g: float = 6.6743e-11) -> NBodyResults:
    return NBodyResults(*simulate_nbody_and_process(system.r_list, 
                                      system.v_list, 
                                      system.m_list, dt, max_time, g))

File: nbodysolver/src/n_body_wrapper.rs
use crate::n_body_core::{process_data_nbody, simulate_nbody};
use ndarray::prelude::*;

pub struct NBodySystem {
    pub r_list: Array2<f64>,
    pub v_list: Array2<f64>,
    pub m_list: Array1<f64>,
}

pub struct NBodyResults {
    pub all_time: Array1<f64>,
    pub all_r: Array3<f64>,
    pub all_v: Array3<f64>,
    pub all_ke: Array1<f64>,
    pub all_pe: Array1<f64>,
    pub all_te: Array1<f64>,
}

pub fn simulate_system(dataset: NBodySystem, dt: f64, max_time: f64, g: f64) -> NBodyResults {
    let (all_time, all_r, all_v, all_ke, all_pe, all_te) = simulate_nbody(
        dataset.r_list,
        dataset.v_list,
        dataset.m_list,
        dt,
        max_time,
        g,
    );
    NBodyResults {
        all_time: Array::from_vec(all_time),
        all_r: process_data_nbody(all_r),
        all_v: process_data_nbody(all_v),
        all_ke: Array::from_vec(all_ke),
        all_pe: Array::from_vec(all_pe),
        all_te: Array::from_vec(all_te),
    }
}
